3rdparty/hal_rvv/hal_rvv_1p0/pyramids.hpp

```diff
@@ -0,0 +1,681 @@
+    return CV_HAL_ERROR_OK;
+}
+
+template<typename T, typename WT>
+void PyrDownInvoker<T, WT>::operator()(const Range& range) const
+{
+    const int PD_SZ = 5;
+
+    int bufstep = (dst_width*cn + 15) & -16;
+    std::vector<WT> _buf(bufstep*PD_SZ + 16);
+    WT* buf = (WT*)(((size_t)_buf.data() + 15) & -16);
+    WT* rows[PD_SZ];
+
+    int sy0 = -PD_SZ/2, sy = range.start * 2 + sy0, width0 = std::min((src_width-PD_SZ/2-1)/2 + 1, dst_width);
+
+    int _dst_width = dst_width * cn;
+    width0 *= cn;
+
+    for (int y = range.start; y < range.end; y++)
+    {
+        T* dst = reinterpret_cast<T*>(dst_data + dst_step * y);
+        WT *row0, *row1, *row2, *row3, *row4;
+
+        // fill the ring buffer (horizontal convolution and decimation)
+        int sy_limit = y*2 + 2;
+        for( ; sy <= sy_limit; sy++ )
+        {
+            WT* row = buf + ((sy - sy0) % PD_SZ)*bufstep;
+            int _sy = borderInterpolate(sy, src_height, borderType);
+            const T* src = reinterpret_cast<const T*>(src_data + src_step * _sy);
+
+            do {
+                int x = 0;
+                for( ; x < cn; x++ )
+                {
+                    row[x] = src[tabL[x+cn*2]]*6 + (src[tabL[x+cn]] + src[tabL[x+cn*3]])*4 +
+                        src[tabL[x]] + src[tabL[x+cn*4]];
+                }
+
+                if( x == _dst_width )
+                    break;
+
+                pyrDownVec0<T, WT>()(src, row, reinterpret_cast<const uint*>(tabM), cn, width0);
+                x = width0;
+
+                // tabR
+                for (int x_ = 0; x < _dst_width; x++, x_++)
+                {
+                    row[x] = src[tabR[x_+cn*2]]*6 + (src[tabR[x_+cn]] + src[tabR[x_+cn*3]])*4 +
+                        src[tabR[x_]] + src[tabR[x_+cn*4]];
+                }
+            } while (0);
+        }
+
+        // do vertical convolution and decimation and write the result to the destination image
+        for (int k = 0; k < PD_SZ; k++)
+            rows[k] = buf + ((y*2 - PD_SZ/2 + k - sy0) % PD_SZ)*bufstep;
+        row0 = rows[0]; row1 = rows[1]; row2 = rows[2]; row3 = rows[3]; row4 = rows[4];
+
+        pyrDownVec1<T, WT>()(row0, row1, row2, row3, row4, dst, _dst_width);
+    }
+}
+
+// the algorithm is copied from imgproc/src/pyramids.cpp,
+// in the function template void cv::pyrUp_
+template<typename T, typename WT>
+inline int pyrUp(const uchar* src_data, size_t src_step, int src_width, int src_height, uchar* dst_data, size_t dst_step, int dst_width, int dst_height, int cn)
+{
+    const int PU_SZ = 3;
+
+    int bufstep = ((dst_width+1)*cn + 15) & -16;
+    std::vector<WT> _buf(bufstep*PU_SZ + 16);
+    WT* buf = (WT*)(((size_t)_buf.data() + 15) & -16);
+    std::vector<int> _dtab(src_width*cn);
+    int* dtab = _dtab.data();
+    WT* rows[PU_SZ];
+
+    CV_Assert( std::abs(dst_width - src_width*2) == dst_width % 2 &&
```

Comment:
Let's avoid `CV_Assert` too. Just return NOT_IMPLEMENTED in this case. Here and in other places.